source/postprocess/particle_density_histogram.cc

```diff
@@ -0,0 +1,268 @@
+    {
+      double  local_min_score = std::numeric_limits<double>::max();
+      double local_max_score = 0;
+      double global_score = 0;
+      unsigned int cells_with_particles = 0;
+
+
+      for (const typename Triangulation<dim>::active_cell_iterator &cell : this->get_dof_handler().active_cell_iterators())
+      {
+        if (cell->is_locally_owned())
+          {
+            /*
+              Create a table with perfect distribution (even number of particles in each bucket)
+              take the distance squared between the "perfect" table and the worst case table
+              Create a table with the actual distribution
+              Take distance between "perfect" table and the actual table
+              return the ratio between the observed distance and the worst case distance.
+              a value of 1 represents the worst possible distribution, a value of 0 represents a completely even distribution.
+            */
+            unsigned int particles_in_cell = 0;
+            for (unsigned int particle_manager_index = 0; particle_manager_index < this->n_particle_managers(); ++particle_manager_index)
+            {
+              particles_in_cell += this->get_particle_manager(particle_manager_index).get_particle_handler().n_particles_in_cell(cell);
+            }
+            if(particles_in_cell > 0)
+            {
+              cells_with_particles++;
+              double particles_in_cell_double = static_cast<double>(particles_in_cell);
+              double granularity_double = static_cast<double>(granularity);
+              double ideal_n_particles_per_bucket = particles_in_cell_double/(granularity_double*granularity_double);
+
+              /*
+              buckets_ideal contains doubles because the ideal
+              number of particles per bucket will not always be an integer.
+              */
+              Table<dim,double> buckets_ideal; 
+              TableIndices<dim> bucket_sizes;
+              for (unsigned int i=0; i<dim; ++i){
+                bucket_sizes[i] = granularity;
+              }
+              buckets_ideal.reinit(bucket_sizes);
+              double bucket_width = 1.0/granularity_double;
+              buckets_ideal.fill(ideal_n_particles_per_bucket);
+
+              Table<dim,unsigned int> buckets_actual; 
+              buckets_actual.reinit(bucket_sizes);
+              sortParticles(cell,buckets_actual,bucket_width);
+              /*
+              in the worst case, all particles are in one bucket. 
+              (granularity*dim)-1 is equal to the number of buckets
+              in the table minus 1 (the bucket all the particles are in)
+              */
+              double worst_case_error_squared = 
+              ((particles_in_cell_double-ideal_n_particles_per_bucket)*(particles_in_cell_double-ideal_n_particles_per_bucket))+
+              ((ideal_n_particles_per_bucket*ideal_n_particles_per_bucket)*((granularity_double*static_cast<double>(dim))-1.0));
```

Comment:
and here as well, wouldnt the number of buckets be granularity to the power of dim?